### PR TITLE
[ASA-398][SM]: correct the auth action's exception handling

### PR DESCRIPTION
**Pull request by sureshhmrc**

### CODE REVIEW

File: app/uk/gov/hmrc/agentsubscription/controllers/SubscriptionController.scala [hunk 1]
```diff
@@ -16,46 +16,61 @@
...
 
 @Singleton
 class SubscriptionController @Inject() (subscriptionService: SubscriptionService)(implicit
   metrics: Metrics,
   microserviceAuthConnector: MicroserviceAuthConnector)
   extends AuthActions(metrics, microserviceAuthConnector) with BaseController {
 
-  def createSubscription = affinityGroupAndEnrolments { implicit request => implicit authIds =>
-    implicit val hc = fromHeadersAndSession(request.headers, None)
-    withJsonBody[SubscriptionRequest] { subscriptionRequest =>
```
Comment (rvijayan) [inline app/uk/gov/hmrc/agentsubscription/controllers/SubscriptionController.scala]: What's wrong with withJsonBody?

File: app/uk/gov/hmrc/agentsubscription/controllers/SubscriptionController.scala [hunk 2]
```diff
@@ -16,46 +16,61 @@
...
+      case Some(JsSuccess(subscriptionRequest, _)) ⇒
+        subscriptionService.createSubscription(subscriptionRequest, authIds).map {
+          case Some(a) => Created(toJson(SubscriptionResponse(a)))
+          case None => Forbidden
+        }.recover {
+          case _: EnrolmentAlreadyAllocated => Conflict
+          case _: IllegalStateException | _: Upstream5xxResponse => InternalServerError
+        }
+
+      case Some(JsError(_)) ⇒
```
Comment (rvijayan) [inline app/uk/gov/hmrc/agentsubscription/controllers/SubscriptionController.scala]: `⇒` is different from what we use in this code base :)

File: app/uk/gov/hmrc/agentsubscription/service/SubscriptionService.scala [hunk 3]
```diff
@@ -23,6 +23,7 @@ import play.api.mvc.Request
 import uk.gov.hmrc.agentmtdidentifiers.model.{ Arn, Utr }
 import uk.gov.hmrc.agentsubscription._
 import uk.gov.hmrc.agentsubscription.audit.{ AgentSubscriptionEvent, AuditService }
+import uk.gov.hmrc.agentsubscription.auth.AuthActions.AuthIds
```
Comment (rvijayan) [inline app/uk/gov/hmrc/agentsubscription/service/SubscriptionService.scala]: unused import?

File: app/uk/gov/hmrc/agentsubscription/controllers/SubscriptionController.scala (hunk 2 above)
Comment (sureshhmrc) [reply app/uk/gov/hmrc/agentsubscription/controllers/SubscriptionController.scala]: yes, its a short form :)

File: app/uk/gov/hmrc/agentsubscription/controllers/SubscriptionController.scala (hunk 1 above)
Comment (sureshhmrc) [reply app/uk/gov/hmrc/agentsubscription/controllers/SubscriptionController.scala]: earlier AuthActions class was dealing with parsing request json which it should not worry about....  `Action.async(parse.json)` 
its not Auth class responsibility to parse the request, so refactored it .. :)

File: app/uk/gov/hmrc/agentsubscription/service/SubscriptionService.scala (hunk 3 above)
Comment (sureshhmrc) [reply app/uk/gov/hmrc/agentsubscription/service/SubscriptionService.scala]: looks like it was removed in another PR , cant see it anymore

**Comment by sureshhmrc** [comment]
@rvijayan is there anything pending in this PR ?

File: app/uk/gov/hmrc/agentsubscription/controllers/SubscriptionController.scala (hunk 2 above)
Comment (rvijayan) [reply app/uk/gov/hmrc/agentsubscription/controllers/SubscriptionController.scala]: withJsonBody comes from  uk.gov.hmrc.play.microservice.controller.BaseController.  It basically deals with JsErrors and other BadRequests. May be, we should keep it as it is and remove `case Some(JsError(_))` and `case None`

Comment (sureshhmrc) [reply app/uk/gov/hmrc/agentsubscription/controllers/SubscriptionController.scala]: ok, fixed

**Pull request merged**